src/kvstore/comm.h

```diff
@@ -635,6 +656,302 @@ class CommDevice : public Comm {
+      CHECK(device_ids_ == dev_ids) << "NCCL KVStore supports only single set of devices";
+      CopyFromTo(src, dst[root_id], priority);
+      if (dst.size() == 1) return;
+      std::vector<Engine::VarHandle> mutable_vars;
+      for (size_t i = 0; i < dst.size(); ++i) {
+        if ( i != root_id)
+          mutable_vars.push_back(dst[i].var());
+      }
+      Engine::Get()->PushSync([src, dst, root_id, this](RunContext rctx) {
+          {
+            std::lock_guard<std::mutex> l(Storage::Get()->GetMutex(Context::kGPU));
+            int root = nccl_data_[src.ctx().dev_id].rank;
+            ncclGroupStart();
+            for (size_t i = 0; i < dst.size(); ++i) {
+              auto& bcast = (i == root_id) ? src : dst[i];
+              NCCLEntry cur = nccl_data_[bcast.ctx().dev_id];
+              MSHADOW_TYPE_SWITCH(bcast.dtype(), DType,
+                  ncclBcast(bcast.data().dptr<DType>(),
+                    bcast.shape().Size(),
+                    GetNCCLType(bcast.dtype()),
+                    root,
+                    cur.comm,
+                    cur.stream););
+            }
+            ncclGroupEnd();
+          }
+      },
+      Context::CPU(),
+      {src.var()},
+      mutable_vars,
+      FnProperty::kCPUPrioritized,
+      priority,
+      PROFILER_MESSAGE("KVStoreBCast"));
+    }
+  }
+
+ private:
+  ncclDataType_t GetNCCLType(int dtype) {
+    switch (dtype) {
+      case mshadow::kFloat32:
+        return ncclFloat;
+      case mshadow::kFloat16:
+        return ncclHalf;
+      case mshadow::kFloat64:
+        return ncclDouble;
+      case mshadow::kUint8:
+        return ncclChar;
+      case mshadow::kInt32:
+        return ncclInt;
+      default:
+        LOG(FATAL) << "Unknown type passed to NCCL KVStore";
+    }
+    return ncclNumTypes;
+  }
+
+  void InitNCCL(const std::vector<Context>& devs) {
+    for (size_t i = 0; i < devs.size(); ++i) {
+      device_ids_.push_back(devs[i].dev_id);
+    }
+    std::sort(device_ids_.begin(), device_ids_.end());
+    std::lock_guard<std::mutex> l(Storage::Get()->GetMutex(Context::kGPU));
+    std::vector<ncclComm_t> comms(devs.size());
+    ncclCommInitAll(&(comms[0]), devs.size(), &(device_ids_[0]));
+    for (size_t i = 0; i < devs.size(); ++i) {
+      NCCLEntry e;
+      e.dev_id = device_ids_[i];
+      e.comm = comms[i];
+      e.rank = i;
+      cudaSetDevice(e.dev_id);
+      cudaStreamCreate(&(e.stream));
+      nccl_data_[device_ids_[i]] = e;
+    }
+  }
+
+  using KeyAttrs = std::tuple<int, TShape, int>;
+  // try to allocate buff on device evenly
+  void InitMergeBuffer(const std::vector<Context>& devs) {
+    for (size_t i = 0; i < sorted_key_attrs_.size(); ++i) {
+      int key  = std::get<0>(sorted_key_attrs_[i]);
+      TShape s = std::get<1>(sorted_key_attrs_[i]);
+      int type = std::get<2>(sorted_key_attrs_[i]);
+      auto& buf = merge_buf_[key];
+      Context ctx;
+      // use devs[0] as root
+      ctx = devs[0];
```

Comment:
I will remove the comment.
We want to use `devs[0]` every time because this helps hide some of the latencies and keep more inter-GPU links occupied if the flow of data is always the same.